File: 2016/day15.cpp
// https://adventofcode.com/2016/day/15

#include "../common/common.h"
#include "../common/rust.h"

#ifndef AOC_IMPORT_STD
#include <algorithm>
#include <print>
#endif

struct Disc {
  usize num_pos;
  usize pos;
};

using Input = Vec<Disc>;

/// The disc that part 2 adds below the input's discs
constexpr Disc EXTRA_DISC_VALUE{
    .num_pos = 11,
    .pos = 0,
};

auto parse(String const& filename) -> Input {
  return aoc::views::read_lines(filename) |
         stdv::transform([](str line) {
           // Disc #D has N positions; at time=0, it is at position P.
           let words = aoc::split(line, ' ');
           return Disc{
               .num_pos = aoc::to_number<usize>(words[3]),
               .pos = aoc::to_number<usize>(aoc::rtrim(words[11], ".")),
           };
         }) |
         aoc::collect_vec<Disc>();
}

template <bool EXTRA_DISC>
fn solve_case(Input discs) -> usize {
  if constexpr (EXTRA_DISC) {
    // Part 2 adds one more disc below the ones from the input
    discs.push_back(EXTRA_DISC_VALUE);
  }
  // The first time to press the button so that the capsule
  // falls through every disc's slot
  return *stdr::find_if(Range{0uz}, [&](usize time) {
    // The capsule reaches disc `id` one second per disc after the press
    return stdr::all_of(discs | stdv::enumerate, [=](let& entry) {
      let[index, disc] = entry;
      return ((time + static_cast<usize>(index) + 1 + disc.pos) %
              disc.num_pos) == 0;
    });
  });
}

int main() {
  std::println("Part 1");
  let example = parse("day15.example");
  AOC_EXPECT_RESULT(5, solve_case<false>(example));
  let input = parse("day15.input");
  AOC_EXPECT_RESULT(317371, solve_case<false>(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(85, solve_case<true>(example));
  AOC_EXPECT_RESULT(2080951, solve_case<true>(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2016/src/day15.rs
use itertools::Itertools;

struct Disc {
    num_pos: usize,
    pos: usize,
}

type Input = Vec<Disc>;

/// The disc that part 2 adds below the input's discs
const EXTRA_DISC_VALUE: Disc = Disc {
    num_pos: 11,
    pos: 0,
};

fn parse(filename: &str) -> Input {
    return aoc::file::read_lines(filename)
        .iter()
        .map(|line| {
            // Disc #D has N positions; at time=0, it is at position P.
            let words = line.split_whitespace().collect_vec();
            Disc {
                num_pos: words[3].parse().unwrap(),
                pos: words[11].trim_end_matches('.').parse().unwrap(),
            }
        })
        .collect_vec();
}

fn solve_case<const EXTRA_DISC: bool>(discs: &Input) -> usize {
    // The first time to press the button so that the capsule
    // falls through every disc's slot
    (0..)
        .find(|time| {
            // The capsule reaches disc `id` one second per disc after the press
            discs
                .iter()
                // Part 2 adds one more disc below the ones from the input
                .chain(EXTRA_DISC.then_some(&EXTRA_DISC_VALUE))
                .enumerate()
                .all(|(index, disc)| ((time + index + 1 + disc.pos) % disc.num_pos) == 0)
        })
        .unwrap()
}

fn main() {
    println!("Part 1");
    let example = parse("day15.example");
    aoc::expect_result!(5, solve_case::<false>(&example));
    let input = parse("day15.input");
    aoc::expect_result!(317371, solve_case::<false>(&input));

    println!("Part 2");
    aoc::expect_result!(85, solve_case::<true>(&example));
    aoc::expect_result!(2080951, solve_case::<true>(&input));
}
